lib/elixir/lib/inspect/algebra.ex

```diff
@@ -371,55 +371,66 @@ defmodule Inspect.Algebra do
-                })
-
-              inspected_struct =
-                doc_struct
-                |> format(opts.width)
-                |> IO.iodata_to_binary()
-
-              inspect_error =
-                Inspect.Error.exception(
-                  exception: caught_exception,
-                  stacktrace: __STACKTRACE__,
-                  inspected_struct: inspected_struct
-                )
-
-              if opts.safe do
-                opts = %{opts | inspect_fun: Inspect.Opts.default_inspect_fun()}
-                Inspect.inspect(inspect_error, opts)
-              else
-                reraise(inspect_error, __STACKTRACE__)
+    # TODO Convert opts.structs to :default | :as_structs | :as_maps in v2
+    cond do
+      opts.structs ->
+        try do
+          fun.(struct, opts)
+        rescue
+          caught_exception ->
+            # Because we try to raise a nice error message in case
+            # we can't inspect a struct, there is a chance the error
+            # message itself relies on the struct being printed, so
+            # we need to trap the inspected messages to guarantee
+            # we won't try to render any failed instruct when building
+            # the error message.
+            if Process.get(:inspect_trap) do
+              Inspect.Map.inspect_as_map(struct, opts)
+            else
+              try do
+                Process.put(:inspect_trap, true)
+
+                {doc_struct, _opts} =
+                  Inspect.Map.inspect_as_map(struct, %{
+                    opts
+                    | syntax_colors: [],
+                      inspect_fun: Inspect.Opts.default_inspect_fun()
+                  })
+
+                inspected_struct =
+                  doc_struct
+                  |> Inspect.Map.inspect_as_map(%{
+                    opts
+                    | syntax_colors: [],
+                      inspect_fun: Inspect.Opts.default_inspect_fun()
+                  })
+                  |> format(opts.width)
+                  |> IO.iodata_to_binary()
+
+                inspect_error =
+                  Inspect.Error.exception(
+                    exception: caught_exception,
+                    stacktrace: __STACKTRACE__,
+                    inspected_struct: inspected_struct
+                  )
+
+                if opts.safe do
+                  opts = %{opts | inspect_fun: Inspect.Opts.default_inspect_fun()}
+                  Inspect.inspect(inspect_error, opts)
+                else
+                  reraise(inspect_error, __STACKTRACE__)
+                end
+              after
+                Process.delete(:inspect_trap)
               end
-            after
-              Process.delete(:inspect_trap)
             end
-          end
-      end
-    else
-      Inspect.Map.inspect_as_map(struct, opts)
+        end
+
+      not opts.structs and Keyword.get(opts.custom_options, :raw_structs) ->
+        Inspect.Any.inspect(struct, opts)
```

Comment:
We could use `Inspect.Map.inspect_as_struct` here, but I wanted to avoid [duplicating the logic](https://github.com/elixir-lang/elixir/blob/b42dd214780f8a7732c8b9a8eaf95e0b565e29e7/lib/elixir/lib/inspect.ex#L665-L677) from `Inspect.Any`.